### ISSUE TITLE
Trying to Integrate Python.Net in ASP.Net App Fails Error: "Attempted to access an unloaded AppDomain"

### ISSUE THREAD

**Issue by gatapia** [pythonnet/pythonnet#155]
Hi All,

Has anyone had any success in running some python code from an ASP.Net (Web.Api) server?  I have a python module I load on application start and I have a controller that periodically communicates with this module.

I load the python module with the code taken from PyImportTest.cs file.  And I periodically call this loaded module from Controllers using module.InvokeMethod(...).AdManagedObject(...).

This all works great in Tests but under webserver conditions (threading I am assuming) this fails with a: "Attempted to access an unloaded AppDomain" error.

Any pointers?  Is this a threading issue? Can I get back onto the thread that kicked off the python module somehow before calling InvokeMethod?

Or is this a bug with PythonNet?

**Comment by denfromufa** [comment]
@gatapia can you provide minimal reproducible example for both conditions: tests and webserver?

**Comment by denfromufa** [comment]
@gatapia possibly related https://github.com/pythonnet/pythonnet/issues/109 may need to use `PythonEngine.BeginAllowThreads()`

**Comment by gatapia** [comment]
Tried BeginAllowThreads/EndAllowThreads and did not help.  Current Symptoms:
- Appears to work sometimes?? (Perhaps after killing w3wp process)
- Second time (sometimes first also) we get AppDomain unloaded exceptions
- Other times it appears to halt (perhaps at AcquireLock)
- Once its locked once or AppDomain error occurs I have to manually kill w3wp process.
- Using log4net is not much use as once an error happens the logger stops logging so hard to get a good feel for what is happening.

Here is my code:

**Initialisation (Application Start)**
```
private bool InitEngine() {
      if (PythonEngine.IsInitialized) return true;
      PythonEngine.Initialize();                  
      pythread = PythonEngine.BeginAllowThreads();
      pylock = PythonEngine.AcquireLock();      

     Type rt = typeof(Runtime);      
      var PyString_FromString = rt.GetMethod("PyString_FromString", BindingFlags.NonPublic | BindingFlags.Static);
      var pyStrPtr = (IntPtr) PyString_FromString.Invoke(null, new object[] { dir.Replace('\\', '/') });
      var PySys_GetObject = rt.GetMethod("PySys_GetObject", BindingFlags.NonPublic | BindingFlags.Static);
      var SysDotPath = (IntPtr) PySys_GetObject.Invoke(null, new object[] {"path"});
      var PyList_Append = rt.GetMethod("PyList_Append", BindingFlags.NonPublic | BindingFlags.Static);
      PyList_Append.Invoke(null, new object[] {SysDotPath, pyStrPtr});            
      PythonEngine.RunSimpleString(String.Format("import os; os.chdir('{0}');", dir.Replace('\\', '/')));

      module = PythonEngine.ImportModule("src." + pymodule);
      module.InvokeMethod("init_py_module", new [] { new PyString(dir) });
      return true;
    }
```

**Periodic Calls From Controller**
`return (double[]) module.InvokeMethod("periodic_method_returns_num_tupple", args).AsManagedObject(typeof(double[]))`

**Dispose / Application End**
```
      PythonEngine.ReleaseLock(pylock);
      PythonEngine.EndAllowThreads(pythread);
      PythonEngine.Shutdown();
```

**Comment by denfromufa** [comment]
Instead of acquiring and releasing the GIL once in your application, why are you not just wrapping your periodic call like this?

```
with (Py.GIL()) {
  double[] result = module.InvokeMethod(
  "periodic_method_returns_num_tupple", args).AsManagedObject(
  typeof(double[]));
}
```

**Comment by denfromufa** [comment]
@gatapia which python version, bitness, .NET version, pythonnet version are you working with? did you try from develop branch? also it is not clear what python module you are calling into, it may have some impact as well.

**Comment by gatapia** [comment]
@denfromufa interesting you mention the module I'm calling.  The module calls c++ wrapper module (xgboost) so it could be that this is adding to the mess of languages/binaries/etc.

Let me do more experimentation once I get back in front of my dev box.  For now:
I am using:
.Net 4
64 bit builds
dev branch of pythonnet
python 2.7
Web framework is actually Nancy not ASP.Net

Also thanks for that `using (Py.GIL())`  tip.  I had no idea that was possible.

**Comment by gatapia** [comment]
I have tried using a dummy python module with no code (i.e. def periodic_method_returns_num_tupple(): return (0,0)) and it still "halts" after a few calls.  Once it locks up the only way to revive it is to kill w3wp.

**Comment by denfromufa** [comment]
did you try wrapping period calls with  `using (Py.GIL())`  ?

**Comment by denfromufa** [comment]
do you have multiple threads in your app?

**Comment by denfromufa** [comment]
since nancy is open-source can you provide minimal example how to reproduce the full pythonnet - web server interaction?

**Comment by gatapia** [comment]
Yes I've wrapped all calls to python in using(Py.GIL()), multiple threads? Yes, IIS assigns threads to each request so very possible that I am switching between threads between initialisation and consecutive calls to python.  I'll try to mock a project up today and attach to this discussion

**Comment by filmor** [comment]
Since there has been no further activity on this, I'll close the ticket. Feel free to comment again with an example that demonstrates the problem and I'll open it again.
